src/connection.rs

```diff
@@ -0,0 +1,673 @@
+    }
+}
+
+impl fmt::Debug for SendStream {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("SendStream").finish_non_exhaustive()
+    }
+}
+
+/// The receiving API for a bidirectional QUIC stream.
+pub struct RecvStream {
+    inner: quinn::RecvStream,
+}
+
+impl RecvStream {
+    fn new(inner: quinn::RecvStream) -> Self {
+        Self { inner }
+    }
+
+    /// Get the next message sent by the peer over this stream.
+    pub async fn next(&mut self) -> Result<Bytes, RecvError> {
+        match self.next_wire_msg().await? {
+            Some(WireMsg::UserMsg(msg)) => Ok(msg),
+            msg => Err(SerializationError::unexpected(msg).into()),
+        }
+    }
+
+    pub(crate) async fn next_wire_msg(&mut self) -> Result<Option<WireMsg>, RecvError> {
+        WireMsg::read_from_stream(&mut self.inner).await
+    }
+}
+
+impl fmt::Debug for RecvStream {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("RecvStream").finish_non_exhaustive()
+    }
+}
+
+/// The receiving API for a connection.
+pub(crate) struct ConnectionIncoming {
+    message_rx: mpsc::Receiver<Result<Bytes, RecvError>>,
+    _alive_tx: Arc<watch::Sender<()>>,
+}
+
+impl ConnectionIncoming {
+    fn new(
+        endpoint: quinn::Endpoint,
+        peer_addr: SocketAddr,
+        uni_streams: quinn::IncomingUniStreams,
+        bi_streams: quinn::IncomingBiStreams,
+        alive_tx: Arc<watch::Sender<()>>,
+        alive_rx: watch::Receiver<()>,
+    ) -> Self {
+        let (message_tx, message_rx) = mpsc::channel(INCOMING_MESSAGE_BUFFER_LEN);
+
+        // offload the actual message handling to a background task - the task will exit when
+        // `alive_tx` is dropped, which would be when both sides of the connection are dropped.
+        start_message_listeners(
+            endpoint,
+            peer_addr,
+            uni_streams,
+            bi_streams,
+            alive_rx,
+            message_tx,
+        );
+
+        Self {
+            message_rx,
+            _alive_tx: alive_tx,
+        }
+    }
+
+    /// Get the next message sent by the peer, over any stream.
+    pub(crate) async fn next(&mut self) -> Result<Option<Bytes>, RecvError> {
+        self.message_rx.recv().await.transpose()
+    }
+}
+
+fn start_message_listeners(
+    endpoint: quinn::Endpoint,
+    peer_addr: SocketAddr,
+    uni_streams: quinn::IncomingUniStreams,
+    bi_streams: quinn::IncomingBiStreams,
+    alive_rx: watch::Receiver<()>,
+    message_tx: mpsc::Sender<Result<Bytes, RecvError>>,
+) {
+    let handle = tokio::spawn(listen_on_uni_streams(
+        peer_addr,
+        uni_streams,
+        alive_rx.clone(),
+        message_tx.clone(),
+    ));
+    drop(handle); // we don't care what happens with the task
+
+    let handle = tokio::spawn(listen_on_bi_streams(
+        endpoint, peer_addr, bi_streams, alive_rx, message_tx,
+    ));
+    drop(handle); // we don't care what happens with the task
```

Comment:
Done 👍

Reply:
I was of the opinion that we wanted to remove instances of `let _ = ...` altogether. That's why I suggested `let _handle = ...` 
Just in case we want to add a CI check that denys the use of `let _ = ...` in the code later on.
Approving and merging for now. We can address this later if required :)